lua/telescope/previewers/buffer_previewer.lua

```diff
@@ -9,12 +10,97 @@ local pfiletype = require "plenary.filetype"
 local pscan = require "plenary.scandir"
 
 local buf_delete = utils.buf_delete
-local defaulter = utils.make_default_callable
 
 local previewers = {}
 
 local ns_previewer = vim.api.nvim_create_namespace "telescope.previewers"
 
+local has_file = 1 == vim.fn.executable "file"
+
+local function capture(cmd, raw)
+  local f = assert(io.popen(cmd, "r"))
+  local s = assert(f:read "*a")
+  f:close()
+  if raw then
+    return s
+  end
+  s = string.gsub(s, "^%s+", "")
+  s = string.gsub(s, "%s+$", "")
+  s = string.gsub(s, "[\n\r]+", " ")
+  return s
+end
```

Comment:
This is the only thing that looks kinda unexpected. Everything else looks good to me.

I dont get this capture function. First i dont get the reasoning here. https://github.com/nvim-telescope/telescope.nvim/pull/1231#discussion_r710400211 Why wouldnt it work. It executes a command synchronous similar to this function but it uses plenary.job which is a wrapper around [vim.loop.spawn](https://github.com/luvit/luv/blob/master/docs.md#uvspawnpath-options-on_exit). plenary.job has a sync interface. 

Second i would prefer if we use a async version, that isnt blocking because afaik, io.popen is blocking. Plenary.job just abstracts the pain away from using `uv.spawn` but gives us the callback hell. We havent really done a good uv.spawn abstraction that uses plenary.async that we could use here. Also i guess that would require us to rewrite everything :laughing: 

Otherwise looks amazing :) Thanks you very much

Reply:
As soon as `utils.get_os_command_output` is within `vim.loop.fs_stat`, it triggers `SIGABRT` because `Job:sync` relies on `vim.wait` which causes the error signal. I'm not sure, I guess that wouldn't happen if plenary used uv timers?

> Second i would prefer if we use a async version, that isnt blocking because afaik, io.popen is blocking

Yeah, it's blocking. Takes ~ 8ms (`job:sync` is c. 20% faster) for a simple lua file.

> prefer if we use a async version [...] Also i guess that would require us to rewrite everything

Just to check, you mean wrap essentially everything in Job:start for `file` that runs the entire logic inside it? That gets a bit funny though if the user doesn't have file and induces ever more callback hell. Wouldn't it be a good compromise to use `utils.get_os_command_output` outside `vim.loop.fs_stat`. We would only be "blocking" on filetypes plenary doesn't catch anyways.

Reply:
i forgot we are also in a `vim.loop.fs_stat`. Yeah lets merge. Its good enough. We can fix it later. By getting rid of the callback hell and replacing it with `plenary.async`